Refactor text function and integrate another variant of customfonts

arasabbasi opened the pull request.

arasabbasi commented:
@MrRio 
Is this better to accept as merge?

arasabbasi commented:
@MrRio 
I dont know if you checked the code, but it supports align justify and multiline processing.

I strongly recommend to use this PR to implement custom fonts etc.

mm2432 commented:
When will this be merged to master?

arasabbasi commented:
@saurabhgis

For using  the arabic-font support you would have to use atleast 

/jspdf.js
/plugins/ttffont.js
/plugins/ttffont.js
/plugins/ttfsupport.js
/plugins/utf8.js
/plugins/arabic.js

from this PR

```
       var NafeesNastaleeq = ''; //put here the base64 Value of your font
    	var doc = new jsPDF();
	doc.addFileToVFS("NafeesNastaleeq.ttf", NafeesNastaleeq );
	doc.addFont('NafeesNastaleeq.ttf', 'NafeesNastaleeq ', 'normal', 'Identity-H');
	doc.viewerPreferences({Direction: 'R2L'});
        doc.setFont('NafeesNastaleeq '); // set font
        doc.setFontSize(45);
        doc.setTextColor(153, 051, 102);
    	
    	doc.text(['لا إله إلا الله محمد رسول الله','لا إله إلا الله', 'a'], 100, 100, {lang: 'ur})
    	doc.save("test.pdf");
```

You can convert the font file to base64 e.g. on http://www.motobit.com/util/base64-decoder-encoder.asp

You can see the changes in the file tab

https://github.com/MrRio/jsPDF/pull/1448/files

If you want to download all together
https://github.com/arasabbasi/jsPDF/archive/refactor_text.zip

But be aware, that i did not build a jspdf.debug.js or jspdf.min.js

saurabhgis commented:
@arasabbasi Thanks to update jsPDF for arabic & other fonts support. Now I want to know that what to do for Arabic text bcoz we want to use this arabic text(رائحة) in pdf without any custom fonts.
 var doc = new jsPDF();
doc.text(20, 20, 'رائحة');

**Hope you already tested, pls share how to use your code and how to set my custom fonts with jsPDF?**

arasabbasi commented:
@saurabhgis 
 The 14 standard fonts of PDF have only ASCII support. So if you try to use arabic font without custom fonts, then I have to: not possible.

lq0910 commented:
This does not apply to fromhtml()??

lq0910 commented:
@arasabbasi

arasabbasi commented:
@lq0910 
fromHTML uses the standard text-method. But the problem is. that fromHTML is not checking if there is a custom font named XYZ. It uses an internal FontTable and always uses the standard fonts. You have to extend fromHTML to use custom fonts.

lq0910 commented:
@arasabbasi  `var pdf = new jsPDF('p', 'pt', 'letter')
        //pdf.addFileToVFS('STXihei.ttf','STXihei');
        pdf.addFont('STXihei');
        pdf.viewerPreferences({ Direction: 'R2L' });
        pdf.setFont("STXihei");
        pdf.setFontSize(60);

        // pdf.text(15, 45, '早上好。 很高兴见到你');"editor.fontLigatures": true
        // pdf.save('custom_fonts.pdf');
        var margins
            // source can be HTML-formatted string, or a reference
            // to an actual DOM element from which the text will be scraped.
            , source = dom

            //, fonts = [['STXihei', 'SimSun'], ['Microsoft YaHei', '']]
            // we support special element handlers. Register them with jQuery-style 
            // ID selector for either ID or node name. ("#iAmID", "div", "span" etc.)
            // There is no support for any other type of selectors 
            // (class, of compound) at this time.
            , specialElementHandlers = {
                // element with id of "bypass" - jQuery style selector
                '#bypassme': function (element, renderer) {
                    // true = "handled elsewhere, bypass text extraction"
                    return true
                },
                '.hide': function (element, renderer) {
                    // true = "handled elsewhere, bypass text extraction"
                    return true
                }
            }

        margins = {
            top: 80,
            bottom: 60,
            left: 40,
            width: 522
        };

        // all coords and widths are in jsPDF instance's declared units
        // 'inches' in this case

        pdf.fromHTML(
            source // HTML string or DOM elem ref.
            , margins.left // x coord
            , margins.top // y coord
            , {
                'width': margins.width // max width of content on PDF
                , 'elementHandlers': specialElementHandlers
            },
            function (dispose) {
                // dispose: object with X, Y of the last line add to the PDF 
                //          this allow the insertion of new lines after html
                //pdf.autoPrint()
                pdf.save('000000000000.pdf');
            },
            margins
        )`

lq0910 commented:
@arasabbasi  `var pdf = new jsPDF('p', 'pt', 'letter')
        //pdf.addFileToVFS('STXihei.ttf','STXihei');
        pdf.addFont('STXihei');
        pdf.viewerPreferences({ Direction: 'R2L' });
        pdf.setFont("STXihei");
        pdf.setFontSize(60);

        // pdf.text(15, 45, '早上好。 很高兴见到你');"editor.fontLigatures": true
        // pdf.save('custom_fonts.pdf');
        var margins
            // source can be HTML-formatted string, or a reference
            // to an actual DOM element from which the text will be scraped.
            , source = dom

            //, fonts = [['STXihei', 'SimSun'], ['Microsoft YaHei', '']]
            // we support special element handlers. Register them with jQuery-style 
            // ID selector for either ID or node name. ("#iAmID", "div", "span" etc.)
            // There is no support for any other type of selectors 
            // (class, of compound) at this time.
            , specialElementHandlers = {
                // element with id of "bypass" - jQuery style selector
                '#bypassme': function (element, renderer) {
                    // true = "handled elsewhere, bypass text extraction"
                    return true
                },
                '.hide': function (element, renderer) {
                    // true = "handled elsewhere, bypass text extraction"
                    return true
                }
            }

        margins = {
            top: 80,
            bottom: 60,
            left: 40,
            width: 522
        };

        // all coords and widths are in jsPDF instance's declared units
        // 'inches' in this case

        pdf.fromHTML(
            source // HTML string or DOM elem ref.
            , margins.left // x coord
            , margins.top // y coord
            , {
                'width': margins.width // max width of content on PDF
                , 'elementHandlers': specialElementHandlers
            },
            function (dispose) {
                // dispose: object with X, Y of the last line add to the PDF 
                //          this allow the insertion of new lines after html
                //pdf.autoPrint()
                pdf.save('000000000000.pdf');
            },
            margins
        )`

lq0910 commented:
@arasabbasi  ` var pdf = new jsPDF('p', 'pt', 'letter')
        pdf.addFileToVFS('STXihei.ttf','STXihei');
        pdf.addFont('STXihei');
        pdf.viewerPreferences({ Direction: 'R2L' });
        pdf.setFont("STXihei");
        pdf.setFontSize(60);

        // pdf.text(15, 45, '早上好。 很高兴见到你');"editor.fontLigatures": true
        // pdf.save('custom_fonts.pdf');
        var margins
            // source can be HTML-formatted string, or a reference
            // to an actual DOM element from which the text will be scraped.
            , source = dom

            //, fonts = [['STXihei', 'SimSun'], ['Microsoft YaHei', '']]
            // we support special element handlers. Register them with jQuery-style 
            // ID selector for either ID or node name. ("#iAmID", "div", "span" etc.)
            // There is no support for any other type of selectors 
            // (class, of compound) at this time.
            , specialElementHandlers = {
                // element with id of "bypass" - jQuery style selector
                '#bypassme': function (element, renderer) {
                    // true = "handled elsewhere, bypass text extraction"
                    return true
                },
                '.hide': function (element, renderer) {
                    // true = "handled elsewhere, bypass text extraction"
                    return true
                }
            }

        margins = {
            top: 80,
            bottom: 60,
            left: 40,
            width: 522
        };

        // all coords and widths are in jsPDF instance's declared units
        // 'inches' in this case

        pdf.fromHTML(
            source // HTML string or DOM elem ref.
            , margins.left // x coord
            , margins.top // y coord
            , {
                'width': margins.width // max width of content on PDF
                , 'elementHandlers': specialElementHandlers
            },
            function (dispose) {
                // dispose: object with X, Y of the last line add to the PDF 
                //          this allow the insertion of new lines after html
                //pdf.autoPrint()
                pdf.save('000000000000.pdf');
            },
            margins
        )`

lq0910 commented:
![image](https://user-images.githubusercontent.com/19302074/35431455-85ac5566-02b7-11e8-9d2f-91a276cb04f9.png)

lq0910 commented:
![image](https://user-images.githubusercontent.com/19302074/35431489-a097b9ba-02b7-11e8-8029-644776077728.png)

arasabbasi commented:
And? You have to modify the code in fromHTML

lq0910 commented:
@arasabbasi  yes, Chaos can't be solved.
![image](https://user-images.githubusercontent.com/19302074/35433500-96439f68-02be-11e8-83bb-44981186e229.png)

lq0910 commented:
@arasabbasi 
In fact, I am looking for a solution, I can't modify the source code, just hope to get your help in order to solve the problem.

mgroeneweg commented:
@arasabbasi First of all thanks a lot for all the effort here. However, I have trouble getting your example to work. I added an additional quote in the options of the doc.text call, added plugin vfs.js because it seemed to need it but I cannot get it to work. 

Could you please post another example?

Thanks in advance for your assistance.

With regards,

Marcel

arasabbasi commented:
Yes my friends. You were not the only one, who was strangely surprised by the fact that utf8 didnt work anymore. I was applying the last changes to get a R2L option done. But I figured it out.  When I merged the branch 'feature/custom-fonts' into refactor_text -branch I overwrote some necessary code. I dont know why I did it. But I will untangle this mess...

arasabbasi commented:
Added R2L-Option to reverse the text, important for chinese/japanese/korean writing system
small changes to context2d.js so that it works in IE 11 without warning/error
added tests to PullRequest, with working russian text and arabic text.

minor bugfixes in logic and some small refactoring to reduce repetive code.

@MrRio

mgroeneweg commented:
Thank you for the quick follow up! I will download the new version

arasabbasi commented:
@mgroeneweg 

I dont know how to build the jspdf.debug.js. But I use the attached code to test it. Just unpack it into the existing jsPDF/examples Folder. then open jsPDF/examples/basic.html and press "Run Code"-Button of the "Simple two-page text document"-demo and it should generate a russian pdf. 

[examples.zip](https://github.com/MrRio/jsPDF/files/1718820/examples.zip)

mgroeneweg commented:
@arasabbasi Thanks for all the help.

mgroeneweg commented:
@arasabbasi I can generate a PDF now but it seems I'm missing a configuration item because the Arabic text in my own little test comes out garbled, when opening in Adobe rather than the browser, I got a message that my Amiri-regular.ttf has an invalid /BBox. Do you have any idea what that means and what could be the cause of this?

Thanks in advance,

With regards,

Marcel

arasabbasi commented:
That is strange. I can open it in my Acrobat Reader DC without any problem. Can you supply me your generated pdf and tell me which version of Acrobat Reader you are using? Or do you mean by Adobe that you are using the professional software?

mgroeneweg commented:
I have Acrobat reader 11.23
This is the file:
[arabic.pdf](https://github.com/MrRio/jsPDF/files/1719806/arabic.pdf)

arasabbasi commented:
Hmm, the file is already corrupt.

Did you include the previous in this PR existing plugins/customfonts.js? If yes please remove this js-file and try again. It was from the original customfonts-solution but is messing with "my" solution.

arasabbasi commented:
These are my files. Can you unpack and use the basic.html as described above and tell me if there is still wrong file generation?

mgroeneweg commented:
Did you intend to upload a file with that comment?

arasabbasi commented:
Ups. Forgot the zip-File
[jsPDF.zip](https://github.com/MrRio/jsPDF/files/1719867/jsPDF.zip)

mgroeneweg commented:
Yes I will test right now.

mgroeneweg commented:
I have no clue what it says so I hope it is happy! :-)
![image](https://user-images.githubusercontent.com/6723899/36149938-0920b5e6-10c2-11e8-8759-305bd5b6cdd7.png)
So it is somewhere in my configuration.
Does this version also have an example in Arabic?

Thanks for all the help! Really appreciate it!

arasabbasi commented:
See: https://github.com/arasabbasi/jsPDF/blob/refactor_text/tests/unicode/arabic.spec.js

It is necessary to use {lang: 'ar'} as options so that the arabic parser gets activated. But be aware, that I used the arabic letter table from tpdf or mpdf project (dont know which one) and wrote the parser new. So highly probable that arabic parsing is in some kind incorrect. Probably if it is used in wide field, it will show some more "displaying" errors.

mgroeneweg commented:
Yes I noticed that one, but I found it difficult to see how to execute it directly. At least my little test looks similar except the save, which uses a 'unicode' parameter, which I don't know how to pass using doc.save. I was suspicious about the save because everything else seems to work just fine.

In my real project i use doc.output("blob") to get the data. Do I need more parameters for the unicode to export correctly? I find the official documentation a little lacking here.

As soon as I get it working, I will ask the Arab speaking customer colleague to testdrive it.

Thanks a lot!

arasabbasi commented:
You know... better a good and extendable arabic parser than one which is not existing or not easily fixable ;). So if you find arabic errors, please report them. I am not a native arabic speaker, so I have no "chance" to find them all by myself. I was just happy, that I could write basic arabic texts into a pdf.

You simply copy the code in the test and paste it into your basic.js where you have the russian code. Instead of the "comparePDF" you use as usual doc.save("bla.pdf"). But be aware, that looking the test file in the browser is wrong... 
![image](https://user-images.githubusercontent.com/5059100/36151543-e9e57044-10c7-11e8-8869-3ea7d3b30eac.png)
![image](https://user-images.githubusercontent.com/5059100/36151583-0d6804be-10c8-11e8-84af-8ea8bdeed489.png)

mgroeneweg commented:
I'm really happy that you put all the effort in. I will report any text errors found by the testers

arasabbasi commented:
There are some minor problems maybe. Read #1612. 

There I wrote a code, which should remove the unicode high byte, but the issuer found another solution. 
This could be also solved e.g. by using filters, which are not implemented yet, like #1493, but would make the file bigger too.

mgroeneweg commented:
When I do a copy-paste of the Arabic test case into basic.js it does generate a nice Arabic PDF. So something got messed up in my test project.

I will integrate this into the real thing as soon as I have completed my current task. Hopefully tomorrow but definitely Thursday. I will keep you posted.

mgroeneweg commented:
BTW what is your experience with this setup and addImage?

MrRio commented:
This is looking really good. There's some code style bits, but I can sort those post-merge.

mgroeneweg commented:
@MrRio Will you be making a release with this pull request in it? That would be great!

arasabbasi commented:
- added test for japanese kana
- removed examples from previous custom fonts
- added example into basic.html, basic.js (will not wirk till new jspdf.debug.js with custom fonts is published)

MugilarasuCS commented:
Hi 

I am facing issue with arabic content 1. RTL and 2. langauage while wrting arabic content shows some values . 

Please can you help me with example for both the issues.

arasabbasi commented:
Hello, 

usually others come up with non-working examples and not otherwise. 

I think you can find everything here...

shujaatxp commented:
I am having below issue. urgent help is required

<script type="text/javascript" src="~/Scripts/jspdf/html2canvas.js"></script>
<script src="~/Scripts/jspdf/jspdf.min.js"></script>
<script src="~/Scripts/jspdf/plugins/vfs.js"></script>
<script src="~/Scripts/jspdf/libs/ttffont.js"></script>
<script src="~/Scripts/jspdf/plugins/ttfsupport.js"></script>
<script src="~/Scripts/jspdf/plugins/viewerpreferences.js"></script>
<script src="~/Scripts/jspdf/plugins/utf8.js"></script>
<script src="~/Scripts/jspdf/plugins/arabic.js"></script>
<script src="~/Scripts/jspdf/plugins/from_html.js"></script>


  var pdf = new jsPDF('p', 'pt', 'a5');

        pdf.addHTML(document.getElementById('certificate'), function () {

            var NafeesNastaleeq = 'TmFmZWVzTmFzdGFsZWVx'; //put here the base64 Value of your font
           
            pdf.addFileToVFS("NafeesNastaleeq.ttf", NafeesNastaleeq);
            pdf.addFont('NafeesNastaleeq.ttf', 'NafeesNastaleeq ', 'normal', 'Identity-H');
            pdf.viewerPreferences({ Direction: 'R2L' });
            pdf.setFont('NafeesNastaleeq '); // set font
            pdf.setFontSize(45);
            pdf.setTextColor(153, 051, 102);
    	
            pdf.text(['لا إله إلا الله محمد رسول الله', 'لا إله إلا الله', 'a'], 100, 100, { lang: 'ar' });
           
            pdf.internal.scaleFactor = 3.75;
            pdf.save('@Model.FullName-@DateTime.Now.Ticks' + '.pdf');


<img width="427" alt="arabic" src="https://user-images.githubusercontent.com/36654782/36408248-be4f0f48-161d-11e8-8e00-e9531045faab.PNG">

arasabbasi commented:
Will not work, because addHTML doesnt support the customfonts and how should it know that it is arabic text?

arasabbasi commented:
Please dont use this thread for support requests. This here is actually a pull request. 

What you want is using addHTML.js with custom fonts. How you call the stuff is already very wrong. And secondly addHTML doesnt support custom fonts. So what you want to do doesnt work anyway.
